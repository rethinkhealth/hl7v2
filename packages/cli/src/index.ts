#!/usr/bin/env node
/** biome-ignore-all lint/suspicious/noConsole: This is a CLI */
import { readFileSync } from 'node:fs';
import { createInterface } from 'node:readline';

import { type HL7v2Node, parseHL7 } from '@rethinkhealth/hl7v2';
import { Command } from 'commander';

const program = new Command();

program.name('hl7').description('HL7v2 CLI Tool').version('0.1.0');

program
  .command('parse')
  .description('Parse an HL7v2 message')
  .argument('[file]', 'Path to HL7 file')
  .action(async (file) => {
    let message: string;

    if (file) {
      message = readFileSync(file, 'utf8');
    } else {
      // Interactive mode
      const rl = createInterface({
        input: process.stdin,
        output: process.stdout,
      });
      message = await new Promise<string>((resolve) =>
        rl.question('Paste HL7 message:\n', resolve)
      );
      rl.close();
    }

    const dom = parseHL7(message);

    // Show parsed tree
    console.log('\nParsed HL7 DOM:\n');
    console.log(JSON.stringify(dom, null, 2));

    // Basic stats
    const segments = dom.children?.length || 0;
    console.log('\nStats:');
    console.log(`- Segments: ${segments}`);
    console.log(`- Characters: ${message.length}`);
    console.log(`- Message Type: ${getMessageType(dom)}`);
  });

program.parse(process.argv);

// Extract MSH-9 for message type
function getMessageType(dom: HL7v2Node): string {
  const msh = dom.children?.find((n: HL7v2Node) => n.name === 'MSH');
  return msh?.children?.[8]?.value || 'Unknown';
}
