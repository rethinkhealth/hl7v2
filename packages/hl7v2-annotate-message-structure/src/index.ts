import type { Root, RootData } from "@rethinkhealth/hl7v2-ast";
import type { MessageInfo } from "@rethinkhealth/hl7v2-util-message-info";
import type { Plugin } from "unified";

/**
 * Unified plugin to infer HL7v2 message structure when missing.
 *
 * This plugin infers the message structure (MSH-9.3) from message code (MSH-9.1)
 * and trigger event (MSH-9.2) when the structure field is not present in the source.
 *
 * The inferred structure follows the convention: `{messageCode}_{triggerEvent}`
 *
 * @example
 * ```typescript
 * import { unified } from 'unified';
 * import { hl7v2Parser } from '@rethinkhealth/hl7v2-parser';
 * import { hl7v2AnnotateMessage } from '@rethinkhealth/hl7v2-annotate-message';
 * import { hl7v2AnnotateMessageStructure } from '@rethinkhealth/hl7v2-annotate-message-structure';
 *
 * const processor = unified()
 *   .use(hl7v2Parser)
 *   .use(hl7v2AnnotateMessage)
 *   .use(hl7v2AnnotateMessageStructure);
 *
 * const tree = processor.parse(message);
 * await processor.run(tree);
 *
 * console.log(tree.data.messageInfo);
 * // { version: "2.5", messageCode: "ADT", triggerEvent: "A01", messageStructure: "ADT_A01" }
 * ```
 */
export const hl7v2AnnotateMessageStructure: Plugin<[], Root, Root> = () => {
  return (tree: Root) => {
    // Ensure data object exists
    if (!tree.data) {
      return tree;
    }

    // Get messageInfo from tree.data (cast to include messageInfo)
    const messageInfo = (tree.data as RootData & { messageInfo?: MessageInfo })
      .messageInfo;

    // If messageInfo doesn't exist, nothing to do
    if (!messageInfo) {
      return tree;
    }

    // If messageStructure is already set, don't override it
    if (messageInfo.messageStructure) {
      return tree;
    }

    // If we have both messageCode and triggerEvent, infer the structure
    if (messageInfo.messageCode && messageInfo.triggerEvent) {
      messageInfo.messageStructure = `${messageInfo.messageCode}_${messageInfo.triggerEvent}`;
    }

    return tree;
  };
};

export default hl7v2AnnotateMessageStructure;
