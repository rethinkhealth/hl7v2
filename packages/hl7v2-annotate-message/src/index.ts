import type { Root } from "@rethinkhealth/hl7v2-ast";
import type { MessageInfo } from "@rethinkhealth/hl7v2-util-message-info";
import { getMessageInfo } from "@rethinkhealth/hl7v2-util-message-info";
import type { Plugin } from "unified";

/**
 * Augment the RootData type to include messageInfo
 */
declare module "@rethinkhealth/hl7v2-ast" {
  // biome-ignore lint/nursery/useConsistentTypeDefinitions: module augmentation
  export interface RootData {
    /**
     * Message metadata extracted from MSH segment
     * Includes version (MSH-12), message code (MSH-9.1),
     * trigger event (MSH-9.2), and message structure (MSH-9.3)
     */
    messageInfo?: MessageInfo;
  }
}

/**
 * Unified plugin to annotate HL7v2 AST with message metadata.
 *
 * This plugin extracts message metadata from the MSH segment and stores it
 * in `Root.data.messageInfo` for downstream plugins to access.
 *
 * The metadata includes:
 * - `version`: MSH-12 (e.g., "2.5")
 * - `messageCode`: MSH-9.1 (e.g., "ADT")
 * - `triggerEvent`: MSH-9.2 (e.g., "A01")
 * - `messageStructure`: MSH-9.3 (e.g., "ADT_A01")
 *
 * @example
 * ```typescript
 * import { unified } from 'unified';
 * import { hl7v2Parser } from '@rethinkhealth/hl7v2-parser';
 * import { hl7v2AnnotateMessage } from '@rethinkhealth/hl7v2-annotate-message';
 *
 * const processor = unified()
 *   .use(hl7v2Parser)
 *   .use(hl7v2AnnotateMessage);
 *
 * const tree = processor.parse(message);
 * await processor.run(tree);
 *
 * console.log(tree.data.messageInfo);
 * // { version: "2.5", messageCode: "ADT", triggerEvent: "A01", messageStructure: "ADT_A01" }
 * ```
 */
export const hl7v2AnnotateMessage: Plugin<[], Root, Root> = () => {
  return (tree: Root) => {
    const messageInfo = getMessageInfo(tree);

    // Initialize data object if it doesn't exist
    if (!tree.data) {
      tree.data = {};
    }

    // Store message info in Root.data
    tree.data.messageInfo = messageInfo;

    return tree;
  };
};

export default hl7v2AnnotateMessage;
