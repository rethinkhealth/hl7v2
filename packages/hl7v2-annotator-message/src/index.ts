/** biome-ignore-all lint/performance/noBarrelFile: fine */
export * from './plugin';
export type * from './types';
