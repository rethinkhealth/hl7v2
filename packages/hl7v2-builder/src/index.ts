/** biome-ignore-all lint/style/useUnifiedTypeSignatures: fine */
import type {
  Component,
  Field,
  FieldRepetition,
  Group,
  Root,
  RootContent,
  Segment,
} from "@rethinkhealth/hl7v2-ast";
import { u } from "unist-builder";

export function m(...children: RootContent[]): Root {
  return u("root", children);
}

export function g(name: string, ...children: Array<Segment | Group>): Group {
  const group: Group = u("group", children);
  group.name = name;
  return group;
}

// TODO: Add support for string[]
// export function s(name: string, ...fields: string[]): Segment;
// export function s(name: string, ...fields: string[] | Field[]): Segment {
export function s(name: string, ...fields: Field[]): Segment {
  const segment: Segment = u("segment", fields);
  segment.name = name;
  return segment;
}

type FieldValue = FieldRepetition | Component | string;
type Flattenable<T> = T | T[];

function flatten<T>(values: Flattenable<T>[]): T[] {
  return values.flatMap((value) => (Array.isArray(value) ? value : [value]));
}

function isFieldRepetition(value: FieldValue): value is FieldRepetition {
  return (
    typeof value === "object" &&
    value !== null &&
    value.type === "field-repetition"
  );
}

function isComponent(value: FieldValue): value is Component {
  return (
    typeof value === "object" && value !== null && value.type === "component"
  );
}

/** Empty field */
export function f(): Field;
export function f(...values: Flattenable<FieldValue>[]): Field;
export function f(...values: Flattenable<FieldValue>[]): Field {
  if (values.length === 0) {
    return u("field", [r()]);
  }

  const flat = flatten<FieldValue>(values);
  if (flat.length === 0) {
    return u("field", [r()]);
  }

  const repetitions: FieldRepetition[] = [];
  let pendingComponents: Array<Component | string> = [];

  const flushPending = () => {
    if (pendingComponents.length === 0) {
      return;
    }
    repetitions.push(r(...pendingComponents));
    pendingComponents = [];
  };

  for (const value of flat) {
    if (isFieldRepetition(value)) {
      flushPending();
      repetitions.push(value);
      continue;
    }

    if (isComponent(value) || typeof value === "string") {
      pendingComponents.push(value);
    }
  }

  flushPending();

  if (repetitions.length === 0) {
    return u("field", [r()]);
  }

  return u("field", repetitions);
}

export function r(): FieldRepetition;
export function r(
  ...components: Flattenable<Component | string>[]
): FieldRepetition;
export function r(
  ...components: Flattenable<Component | string>[]
): FieldRepetition {
  if (components.length === 0) {
    return u("field-repetition", [c()]);
  }

  const flat = flatten<Component | string>(components);
  if (flat.length === 0) {
    return u("field-repetition", [c()]);
  }

  return u(
    "field-repetition",
    flat.map((value) => (typeof value === "string" ? c(value) : value))
  );
}

export function c(): Component;
export function c(...values: Flattenable<string>[]): Component;
export function c(...values: Flattenable<string>[]): Component {
  if (values.length === 0) {
    return u("component", [u("subcomponent", "")]);
  }

  const flat = flatten<string>(values);
  if (flat.length === 0) {
    return u("component", [u("subcomponent", "")]);
  }

  return u(
    "component",
    flat.map((subcomponent) => u("subcomponent", subcomponent))
  );
}
