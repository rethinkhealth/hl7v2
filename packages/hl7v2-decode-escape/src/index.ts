import type { Root, Subcomponent } from '@rethinkhealth/hl7v2-ast';
import {
  DEFAULT_DELIMITERS,
  type HL7v2Delimiters,
} from '@rethinkhealth/hl7v2-utils';
import type { Plugin } from 'unified';
import { visit } from 'unist-util-visit';

export interface HL7v2DecodeOptions {
  delimiters?: Partial<HL7v2Delimiters>;
}

/**
 * Unified plugin to decode HL7v2 escape sequences in subcomponent literals.
 *
 * - Decodes \F\, \S\, \T\, \R\, \E\
 * - Decodes \Xdddd\ hex escapes
 * - Handles \.br\ line breaks
 * - Uses delimiters from Root.data.delimiters if available
 */
export const hl7v2DecodeEscapes: Plugin<[HL7v2DecodeOptions?], Root> = (
  options
) => {
  return (tree: Root) => {
    const delimiters =
      (tree.data as { delimiters?: Partial<HL7v2Delimiters> })?.delimiters ||
      options?.delimiters;

    visit(tree, 'subcomponent', (node: Subcomponent) => {
      const raw = node.value;
      node.value = decode(raw, {
        ...DEFAULT_DELIMITERS,
        ...delimiters,
      });
    });
  };
};

/**
 * Decode HL7v2 escape sequences according to HL7 v2.8 spec.
 *
 * @param value - The value to decode.
 * @param d - The delimiters to use.
 * @returns The decoded value.
 */
// biome-ignore lint/complexity/noExcessiveCognitiveComplexity: this function must handle multiple HL7v2 escape cases and is as simple as possible given the requirements
function decode(value: string, d: typeof DEFAULT_DELIMITERS): string {
  if (!value?.includes(d.escape)) {
    return value;
  }

  let decoded = '';
  let i = 0;

  while (i < value.length) {
    if (value[i] === d.escape) {
      const end = value.indexOf(d.escape, i + 1);
      if (end === -1) {
        decoded += value.slice(i); // unterminated escape
        break;
      }

      const code = value.slice(i + 1, end);

      switch (code) {
        case 'F':
          decoded += d.field;
          break;
        case 'S':
          decoded += d.component;
          break;
        case 'R':
          decoded += d.repetition;
          break;
        case 'T':
          decoded += d.subcomponent;
          break;
        case 'E':
          decoded += d.escape;
          break;
        case '.br':
          decoded += '\r'; // or '\n' if you prefer LF
          break;
        case 'H':
        case 'N':
          // Highlight start/end: ignored for now
          break;
        default:
          if (code.startsWith('X') && code.length > 1) {
            decoded += decodeHexSequence(code.slice(1));
          } else {
            // Unknown escape: preserve as-is
            decoded += d.escape + code + d.escape;
          }
          break;
      }

      i = end + 1;
    } else {
      decoded += value[i++];
    }
  }

  return decoded;
}

/**
 * Decode HL7 \Xdddd\ hexadecimal escape sequences into characters.
 */
function decodeHexSequence(hex: string): string {
  let result = '';
  for (let i = 0; i < hex.length; i += 2) {
    const byte = hex.slice(i, i + 2);
    const codePoint = Number.parseInt(byte, 16);
    if (!Number.isNaN(codePoint)) {
      result += String.fromCharCode(codePoint);
    }
  }
  return result;
}
