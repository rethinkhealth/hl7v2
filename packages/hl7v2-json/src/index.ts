/** biome-ignore-all lint/suspicious/noExplicitAny: we should use a JSON type */
import type { HL7v2Node } from '@rethinkhealth/hl7v2-ast';
import type { Plugin } from 'unified';
import type { Node } from 'unist';

export const hl7v2Jsonify: Plugin<[], HL7v2Node, string> = function (): void {
  // biome-ignore lint/complexity/noUselessThisAlias: this is a plugin
  const self = this;

  function compiler(tree: Node): string {
    return JSON.stringify(toJson(tree as HL7v2Node), null, 2);
  }

  self.compiler = compiler;
};

/**
 * Convert an HL7v2Node tree into a simplified JSON representation.
 * Removes unist metadata and keeps only meaningful HL7v2 properties.
 */
export function toJson(node: HL7v2Node): any {
  const { type, name, index, value, delimiter, children } = node;

  const result: Record<string, any> = {
    type,
    ...(name ? { name } : {}),
    ...(index !== undefined ? { index } : {}),
    ...(value !== undefined ? { value } : {}),
    ...(delimiter ? { delimiter } : {}),
  };

  if (children && children.length > 0) {
    result.children = children.map(toJson);
  }

  return result;
}
