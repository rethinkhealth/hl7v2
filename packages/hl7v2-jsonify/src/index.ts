import type { HL7v2Node } from '@rethinkhealth/hl7v2-ast';
import type { Plugin } from 'unified';
import type { Node } from 'unist';
import { visitParents } from 'unist-util-visit-parents';

type SegmentJSON = {
  segment: string;
  fields: (string | string[])[]; // Nested arrays for fields/repetitions/components
};

export const hl7v2Jsonify: Plugin<[], HL7v2Node, string> = function (): void {
  // biome-ignore lint/complexity/noUselessThisAlias: this is a plugin
  const self = this;

  function compiler(tree: Node): string {
    return JSON.stringify(toJson(tree as HL7v2Node), null, 2);
  }

  self.compiler = compiler;
};

export function toJson(root: HL7v2Node): SegmentJSON[] {
  const segments: SegmentJSON[] = [];
  let currentSegment: SegmentJSON | null = null;

  visitParents<HL7v2Node, HL7v2Node>(root, (node, ancestors) => {
    if (node.type === 'segment') {
      currentSegment = { segment: node.name || 'UNKNOWN', fields: [] };
      segments.push(currentSegment);
    }

    if (node.value !== undefined && currentSegment) {
      // Build numeric path for all indices except segment/root
      const path = [...ancestors, node]
        .filter(
          (n) =>
            n.index !== undefined && n.type !== 'segment' && n.type !== 'root'
        )
        .map((n) => {
          if (typeof n.index !== 'number') {
            throw new Error('HL7 AST node is missing a numeric index');
          }
          // Fields: subtract 1 to skip header (index 0) and convert to 0-based array indices
          // Components/subcomponents: use index as-is (no header to skip)
          return n.type === 'field' ? n.index - 1 : n.index;
        });

      setNestedArrayValue(currentSegment.fields, path, node.value);
    }
  });

  return segments;
}

/**
 * Recursively create nested arrays along the path and set the value.
 */
function setNestedArrayValue(
  arr: (string | string[])[],
  path: number[],
  value: string
) {
  let current: (string | string[])[] = arr;
  for (let i = 0; i < path.length - 1; i++) {
    const idx = path[i];

    if (idx === undefined) {
      return;
    }

    if (!current[idx]) {
      current[idx] = [];
    }
    current = current[idx] as (string | string[])[];
  }

  const lastKey = path.at(-1);
  if (lastKey === undefined) {
    return;
  }

  current[lastKey] = value;
}
