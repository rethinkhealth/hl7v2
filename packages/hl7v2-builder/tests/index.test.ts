import { describe, expect, it } from "vitest";
import { c, f, m, r, s } from "../src";

describe("builder", () => {
  describe("root", () => {
    it("should build a simple root", () => {
      const root = m();
      expect(root).toEqual({
        type: "root",
        children: [],
      });
    });

    it("should build a root with a single segment", () => {
      const root = m(s("MSH", f("another value")));
      expect(root).toEqual({
        type: "root",
        children: [
          {
            type: "segment",
            children: [
              { type: "segment-header", value: "MSH" },
              {
                type: "field",
                children: [
                  {
                    type: "field-repetition",
                    children: [
                      {
                        type: "component",
                        children: [
                          { type: "subcomponent", value: "another value" },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      });
    });

    it("should build a root with a multiple segments", () => {
      const root = m(s("MSH", f("another value")), s("PID", f("12345")));
      expect(root).toEqual({
        type: "root",
        children: [
          {
            type: "segment",
            children: [
              { type: "segment-header", value: "MSH" },
              {
                type: "field",
                children: [
                  {
                    type: "field-repetition",
                    children: [
                      {
                        type: "component",
                        children: [
                          { type: "subcomponent", value: "another value" },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
          {
            type: "segment",
            children: [
              { type: "segment-header", value: "PID" },
              {
                type: "field",
                children: [
                  {
                    type: "field-repetition",
                    children: [
                      {
                        type: "component",
                        children: [{ type: "subcomponent", value: "12345" }],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      });
    });
  });

  describe("segment", () => {
    it("should build a segment with a single-value field", () => {
      const segment = s("MSH", f("another value"));

      expect(segment).toEqual({
        type: "segment",
        children: [
          { type: "segment-header", value: "MSH" },
          {
            type: "field",
            children: [
              {
                type: "field-repetition",
                children: [
                  {
                    type: "component",
                    children: [
                      { type: "subcomponent", value: "another value" },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      });
    });

    it("should build a segment with an array of fields", () => {
      const segment = s("MSH", f("another value"), f("another value 2"));
      expect(segment).toEqual({
        type: "segment",
        children: [
          { type: "segment-header", value: "MSH" },
          {
            type: "field",
            children: [
              {
                type: "field-repetition",
                children: [
                  {
                    type: "component",
                    children: [
                      { type: "subcomponent", value: "another value" },
                    ],
                  },
                ],
              },
            ],
          },
          {
            type: "field",
            children: [
              {
                type: "field-repetition",
                children: [
                  {
                    type: "component",
                    children: [
                      { type: "subcomponent", value: "another value 2" },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      });
    });
  });

  describe("field", () => {
    it("should build an empty field", () => {
      const field = f();
      expect(field).toEqual({
        type: "field",
        children: [
          {
            type: "field-repetition",
            children: [
              {
                type: "component",
                children: [{ type: "subcomponent", value: "" }],
              },
            ],
          },
        ],
      });
    });

    it("should build a single value field", () => {
      const field = f("hello world");
      expect(field).toEqual({
        type: "field",
        children: [
          {
            type: "field-repetition",
            children: [
              {
                type: "component",
                children: [{ type: "subcomponent", value: "hello world" }],
              },
            ],
          },
        ],
      });
    });

    it("should build a field with a single component", () => {
      const field = f(c("hello world!"));
      expect(field).toEqual({
        type: "field",
        children: [
          {
            type: "field-repetition",
            children: [
              {
                type: "component",
                children: [{ type: "subcomponent", value: "hello world!" }],
              },
            ],
          },
        ],
      });
    });

    it("should build a field with an array of components", () => {
      const field = f([c("A"), c("B"), c("C")]);

      expect(field).toEqual({
        type: "field",
        children: [
          {
            type: "field-repetition",
            children: [
              {
                type: "component",
                children: [{ type: "subcomponent", value: "A" }],
              },
              {
                type: "component",
                children: [{ type: "subcomponent", value: "B" }],
              },
              {
                type: "component",
                children: [{ type: "subcomponent", value: "C" }],
              },
            ],
          },
        ],
      });
    });

    it("should build a field with an array of repeated components (field repetition)", () => {
      // Given
      const field = f([
        r([c("A"), c(["B.1", "B.2"])]),
        r([c("C")]),
        r([c("D")]),
      ]);

      // Then
      expect(field).toEqual({
        type: "field",
        children: [
          {
            type: "field-repetition",
            children: [
              {
                type: "component",
                children: [{ type: "subcomponent", value: "A" }],
              },
              {
                type: "component",
                children: [
                  { type: "subcomponent", value: "B.1" },
                  { type: "subcomponent", value: "B.2" },
                ],
              },
            ],
          },
          {
            type: "field-repetition",
            children: [
              {
                type: "component",
                children: [{ type: "subcomponent", value: "C" }],
              },
            ],
          },
          {
            type: "field-repetition",
            children: [
              {
                type: "component",
                children: [{ type: "subcomponent", value: "D" }],
              },
            ],
          },
        ],
      });
    });

    it("should build the same results using a single repetition or component(s)", () => {
      const fieldWithRepetition = f(r(c("A"), c(["B.1", "B.2"])));

      const fieldWithComponents = f(c("A"), c(["B.1", "B.2"]));

      expect(fieldWithRepetition).toEqual(fieldWithComponents);
    });
  });

  describe("component", () => {
    it("should build an empty component", () => {
      const component = c();
      expect(component).toEqual({
        type: "component",
        children: [{ type: "subcomponent", value: "" }],
      });
    });

    it("should build a component with a single value", () => {
      const component = c("hello world!");
      expect(component).toEqual({
        type: "component",
        children: [{ type: "subcomponent", value: "hello world!" }],
      });
    });

    it("should build a component with an array of values", () => {
      const component = c(["hello world!", "hello world 2!"]);
      expect(component).toEqual({
        type: "component",
        children: [
          { type: "subcomponent", value: "hello world!" },
          { type: "subcomponent", value: "hello world 2!" },
        ],
      });
    });
  });

  describe("complex scenarios", () => {
    it("should build a segment with an array of fields, each with an array of components", () => {
      const message = m(
        // MSH Segment
        s(
          "MSH",
          f(),
          f(),
          f("1234"),
          f([c(["A", "A.2", "A.3"]), c("B"), c("C")])
        ),
        // PID Segment
        s("PID", f("12345"), f([c("DOE"), c("JOHN")]))
      );

      expect(message).toMatchSnapshot();
    });
  });
});
