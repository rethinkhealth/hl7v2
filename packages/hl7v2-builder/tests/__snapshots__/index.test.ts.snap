// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`builder > complex scenarios > should build a segment with an array of fields, each with an array of components 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "1234",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "A",
                    },
                    {
                      "type": "subcomponent",
                      "value": "A.2",
                    },
                    {
                      "type": "subcomponent",
                      "value": "A.3",
                    },
                  ],
                  "type": "component",
                },
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "B",
                    },
                  ],
                  "type": "component",
                },
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "C",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
      ],
      "name": "MSH",
      "type": "segment",
    },
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "12345",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "DOE",
                    },
                  ],
                  "type": "component",
                },
                {
                  "children": [
                    {
                      "type": "subcomponent",
                      "value": "JOHN",
                    },
                  ],
                  "type": "component",
                },
              ],
              "type": "field-repetition",
            },
          ],
          "type": "field",
        },
      ],
      "name": "PID",
      "type": "segment",
    },
  ],
  "type": "root",
}
`;
